gtsfm/densify/interface_patchmatchnet.py

```diff
@@ -0,0 +1,178 @@
+        self.keys_map = {}
+        for i in range(self.num_images):
+            self.keys_map[self.keys[i]] = i
+        self.images = images
+        self.image_w = images[self.keys[0]].width
+        self.image_h = images[self.keys[0]].height
+
+        self.pairs, self.depth_ranges = self.configure()
+
+    def configure(self) -> Tuple[np.ndarray, np.ndarray]:
+        """Configure pairs and depth_ranges for each view from sfm_result
+
+        Returns:
+            pairs: array of shape (num_images, num_views-1). Each row_id indicates the index of reference view
+                in self.keys, with (num_views-1) values indicating the indices of source views in self.keys
+            depth_ranges: array of shape (num_images, 2). Each row_id indicates the index of reference view
+                in self.keys, with 2 values indicating [min_depth, max_depth]
+        """
+        num_images = self.num_images
+        num_tracks = self.sfm_result.number_tracks()
+
+        pair_scores = np.zeros((num_images, num_images))
+        # initialize the pairwise scores between the same views as negative infinity
+        np.fill_diagonal(pair_scores, -np.inf)
+
+        depth_ranges = np.zeros((num_images, 2))
+        depth_ranges[:, 0] = np.inf
+
+        for i in range(num_tracks):
+            track_i = self.sfm_result.get_track(i)
+            num_measurements_i = track_i.number_measurements()
+            measurements = [track_i.measurement(j) for j in range(num_measurements_i)]
+            coord_world = track_i.point3()
+            for cam_a in range(num_measurements_i):
+                for cam_b in range(cam_a + 1, num_measurements_i):
+                    cam_a_id = measurements[cam_a][0]
+                    cam_b_id = measurements[cam_b][0]
+
+                    # if both cameras are valid cameras
+                    if cam_a_id in self.keys_map and cam_b_id in self.keys_map:
+
+                        key_a_id = self.keys_map[cam_a_id]
+                        key_b_id = self.keys_map[cam_b_id]
+
+                        # calculate track_i's 3D coordinates in the camera pose
+                        coord_cam_a = self.sfm_result.get_camera(cam_a_id).pose().transformTo(coord_world)
+                        coord_cam_b = self.sfm_result.get_camera(cam_b_id).pose().transformTo(coord_world)
+
+                        # calculate score for measurements of track_i in pair views (cam_a, cam_b)
+                        score_a_b = piecewise_gaussian(p_a=coord_cam_a, p_b=coord_cam_b)
+
+                        # sum up pair scores for each track_i
+                        pair_scores[key_a_id, key_b_id] += score_a_b
+                        pair_scores[key_b_id, key_a_id] += score_a_b
+
+                        # update depth ranges
+                        depth_ranges[(key_a_id, key_b_id), 0] = np.minimum(
+                            depth_ranges[(key_a_id, key_b_id), 0], [coord_cam_a[-1], coord_cam_b[-1]]
```

Comment:
maybe just extract the z coordinates of coord_cam_a and coord_cam_b instead of all 3 coordinates.